File: templates/skeleton/nestjs/src/example/example.service.ts
import { Injectable } from '@nestjs/common';

import { Example, ExampleRepository } from './example.repository';

/** Raised when the requested name is already taken — an expected error. */
export class DuplicateExampleNameError extends Error {
  constructor(readonly name_: string) {
    super(`an example named '${name_}' already exists`);
    this.name = 'DuplicateExampleNameError';
  }
}

/** Raised when no example carries the requested identifier. */
export class ExampleNotFoundError extends Error {
  constructor(readonly id: string) {
    super(`no example with id '${id}'`);
    this.name = 'ExampleNotFoundError';
  }
}

/** Business rules for examples. No HTTP type belongs in this class (A3). */
@Injectable()
export class ExampleService {
  constructor(private readonly repository: ExampleRepository) {}

  /**
   * Registers a new example, enforcing name uniqueness.
   *
   * @throws DuplicateExampleNameError when the name is already in use.
   */
  create(name: string): Example {
    if (this.repository.existsByName(name)) {
      throw new DuplicateExampleNameError(name);
    }
    return this.repository.add(name);
  }

  /**
   * Returns one example.
   *
   * @throws ExampleNotFoundError when no example carries this identifier.
   */
  get(id: string): Example {
    const example = this.repository.get(id);
    if (!example) {
      throw new ExampleNotFoundError(id);
    }
    return example;
  }
}

File: templates/skeleton/python-fastapi/src/@@PROJECT_SNAKE@@/services/example_service.py
"""Business rules for examples.

No FastAPI import belongs in this module: the service knows nothing about HTTP
(AGENTS.md A3 — separation of concerns).
"""

from ..repositories.example_repository import Example, ExampleRepository


class DuplicateExampleError(Exception):
    """Raised when the requested name is already taken — an expected error."""

    def __init__(self, name: str) -> None:
        """Record the name that was already taken."""
        super().__init__(f"an example named {name!r} already exists")
        self.name = name


class ExampleNotFoundError(Exception):
    """Raised when no example matches the requested id — an expected error."""

    def __init__(self, identifier: str) -> None:
        """Record the identifier that could not be found."""
        super().__init__(f"no example with id {identifier!r}")
        self.identifier = identifier


class ExampleService:
    """Creates and reads examples, enforcing the uniqueness rule."""

    def __init__(self, repository: ExampleRepository) -> None:
        """Wire the service to its persistence boundary."""
        self._repository = repository

    def create(self, name: str) -> Example:
        """Create an example.

        Args:
            name: the label to register; must not already exist.

        Returns:
            The created example.

        Raises:
            DuplicateExampleError: when the name is already in use.
        """
        if self._repository.exists_by_name(name):
            raise DuplicateExampleError(name)
        return self._repository.add(name)

    def get(self, identifier: str) -> Example:
        """Return one example.

        Raises:
            ExampleNotFoundError: when no example carries this id.
        """
        example = self._repository.get(identifier)
        if example is None:
            raise ExampleNotFoundError(identifier)
        return example
